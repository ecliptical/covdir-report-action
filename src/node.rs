use serde::de::{Deserialize, Deserializer, MapAccess, Visitor};
use std::fmt;
use std::marker::PhantomData;

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Node {
    pub name: String,

    pub lines_covered: usize,

    pub lines_missed: usize,

    pub lines_total: usize,

    pub coverage_percent: f32,

    #[serde(default, deserialize_with = "deserialize_kv_vec")]
    pub children: Vec<(String, Node)>,
}

fn deserialize_kv_vec<'de, D>(deserializer: D) -> Result<Vec<(String, Node)>, D::Error>
where
    D: Deserializer<'de>,
{
    let items = KeyValueVec::deserialize(deserializer)?;
    Ok(items.into())
}

#[derive(Debug)]
struct KeyValueVec<K, V>(Vec<(K, V)>);

impl<K, V> From<KeyValueVec<K, V>> for Vec<(K, V)> {
    fn from(src: KeyValueVec<K, V>) -> Self {
        src.0
    }
}

struct KeyValueVisitor<K, V> {
    marker: PhantomData<fn() -> KeyValueVec<K, V>>,
}

impl<K, V> KeyValueVisitor<K, V> {
    fn new() -> Self {
        KeyValueVisitor {
            marker: PhantomData,
        }
    }
}

impl<'de, K, V> Visitor<'de> for KeyValueVisitor<K, V>
where
    K: Deserialize<'de>,
    V: Deserialize<'de>,
{
    type Value = KeyValueVec<K, V>;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("a map")
    }

    fn visit_map<M>(self, mut access: M) -> Result<Self::Value, M::Error>
    where
        M: MapAccess<'de>,
    {
        let mut items = Vec::with_capacity(access.size_hint().unwrap_or(1));
        while let Some((key, value)) = access.next_entry()? {
            items.push((key, value));
        }

        Ok(KeyValueVec(items))
    }
}

impl<'de, K, V> Deserialize<'de> for KeyValueVec<K, V>
where
    K: Deserialize<'de>,
    V: Deserialize<'de>,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_map(KeyValueVisitor::new())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn deserialize_map() {
        let res = deserialize_kv_vec(json!({
            "a": {
                "name": "a",
                "linesCovered": 12,
                "linesMissed": 34,
                "linesTotal": 46,
                "coveragePercent": 12.34
            },
            "b": {
                "name": "b",
                "linesCovered": 56,
                "linesMissed": 78,
                "linesTotal": 134,
                "coveragePercent": 56.78
            }
        }));

        assert!(res.is_ok());
        let values = res.unwrap();
        assert_eq!(values.get(0).map(|v| v.0.as_ref()), Some("a"));
        assert_eq!(values.get(1).map(|v| v.0.as_ref()), Some("b"));
    }
}
