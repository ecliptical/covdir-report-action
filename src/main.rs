use config::Config;
use config::Environment;
use gumdrop::Options;
use serde::Deserialize;
use std::fs::File;
use std::io::{BufReader, BufWriter, Write};
use std::path::PathBuf;
use tracing::*;
use tracing_subscriber::EnvFilter;

mod node;

use crate::node::Node;

type ExplicitBool = bool;

#[derive(Debug, Options)]
struct Cmd {
    /// Path to input file.
    #[options(required)]
    file: PathBuf,

    /// Path to output file.
    out: PathBuf,

    /// Write job summary.
    #[options(no_short)]
    summary: ExplicitBool,

    /// Report tile.
    #[options(default = "Line coverage")]
    title: String,

    /// Display help.
    help: bool,
}

#[derive(Debug, Deserialize)]
struct Env {
    output: PathBuf,
    step_summary: PathBuf,
}

fn write_output(file: impl Write, root: &Node) -> std::io::Result<()> {
    let mut out = BufWriter::new(file);
    writeln!(&mut out, "lines_covered={}", root.lines_covered)?;
    writeln!(&mut out, "lines_missed={}", root.lines_missed)?;
    writeln!(&mut out, "lines_total={}", root.lines_total)?;
    writeln!(&mut out, "coverage_percent={}", root.coverage_percent)?;
    Ok(())
}

fn write_summary(file: impl Write, root: &Node, title: &str) -> std::io::Result<()> {
    let mut out = BufWriter::new(file);
    writeln!(
        &mut out,
        "| {title}: | {:.0}% |",
        root.coverage_percent.round()
    )?;
    writeln!(&mut out, "|:---|:---|")?;
    writeln!(
        &mut out,
        "| Lines covered: | {} |",
        fmt_number(root.lines_covered)
    )?;
    writeln!(
        &mut out,
        "| Lines missed: | {} |",
        fmt_number(root.lines_missed)
    )?;
    writeln!(
        &mut out,
        "| Total lines: | {} |",
        fmt_number(root.lines_total)
    )?;
    Ok(())
}

fn fmt_number(n: usize) -> String {
    n.to_string()
        .as_bytes()
        .rchunks(3)
        .map(String::from_utf8_lossy)
        .rev()
        .collect::<Vec<_>>()
        .as_slice()
        .join(",")
}

fn main() -> anyhow::Result<()> {
    tracing_subscriber::fmt()
        .with_env_filter(EnvFilter::from_default_env())
        .with_file(true)
        .with_line_number(true)
        .init();

    let opt = Cmd::parse_args_default_or_exit();

    let config = Config::builder()
        .add_source(Environment::with_prefix("GITHUB"))
        .build()?;

    let env: Env = config.try_deserialize()?;

    debug!(env = ?env, opt = ?opt);

    let file = File::open(opt.file)?;
    let reader = BufReader::new(file);

    let root: Node = serde_json::from_reader(reader)?;

    debug!(root = ?root);

    let out = File::create(env.output)?;
    write_output(out, &root)?;

    if !opt.out.as_os_str().is_empty() {
        let file = File::create(opt.out)?;
        write_summary(file, &root, &opt.title)?;
    }

    if opt.summary {
        let file = File::options()
            .create(true)
            .append(true)
            .open(env.step_summary)?;
        write_summary(file, &root, &opt.title)?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;

    #[test]
    fn fmt_number_short() {
        assert_eq!(fmt_number(123), "123");
    }

    #[test]
    fn fmt_number_medium() {
        assert_eq!(fmt_number(1234), "1,234");
    }

    #[test]
    fn fmt_number_long() {
        assert_eq!(fmt_number(1234567890), "1,234,567,890");
    }

    #[test]
    fn write_output_ok() {
        let root = Node {
            name: String::default(),
            lines_covered: 12,
            lines_missed: 34,
            lines_total: 46,
            coverage_percent: 12.34,
            children: Vec::default(),
        };

        let mut buf = Vec::default();

        let res = write_output(&mut buf, &root);

        assert!(res.is_ok());
        assert_snapshot!(String::from_utf8(buf).unwrap_or_default());
    }

    #[test]
    fn write_summary_ok() {
        let root = Node {
            name: String::default(),
            lines_covered: 12,
            lines_missed: 34,
            lines_total: 46,
            coverage_percent: 12.34,
            children: Vec::default(),
        };

        let mut buf = Vec::default();

        let res = write_summary(&mut buf, &root, "Line coverage");

        assert!(res.is_ok());
        assert_snapshot!(String::from_utf8(buf).unwrap_or_default());
    }
}
