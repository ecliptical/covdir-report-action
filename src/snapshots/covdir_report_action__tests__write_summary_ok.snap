---
source: src/main.rs
assertion_line: 184
expression: "String::from_utf8(buf).unwrap_or_default()"
---
| Line coverage: | 12% |
|:---|:---|
| Lines covered: | 12 |
| Lines missed: | 34 |
| Total lines: | 46 |
